---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# CellTyPETool

<!-- badges: start -->
<!-- badges: end -->

##Description
The goal of CellTyPETool is to facilitate research that requires deconvolution of bulk-tissue RNAseq data. The package can be used to generate cell type proportion estimations from bulk-tissue RNAseq using two different validated methods, markerGeneProfile and BRETIGEA.
While other R packages have been published that calculate cell type proportions (McKenzie et al, 2018; Mancarci et al., 2017), they only offer their own analysis and not a comparison between, nor do they offer visualizations between cell-type proportions associations with disease or other important phenotypes.
Additionally, they donâ€™t offer simple way to check which marker genes were encorporated in the cell-type proportion calculation.

##Installation

To install the latest version of the package:

require("devtools")
devtools::install_github("meconsens/CellTyPETool", build_vignettes = TRUE)
library("CellTyPETool")


## Example

This is a basic example which shows you how to solve a common problem:

```{r example}
library(CellTyPETool)
## basic example code
```

##Acknowledgements
This package was developed as part of an assessment for 2020BCB410H: Applied Bioinformatics, University of Toronto, Toronto,CANADA.


```{r pressure, echo = FALSE}
plot(pressure)
```

